tensorflow_addons/optimizers/discriminative_layer_training.py

```diff
@@ -0,0 +1,146 @@
+        optimizer layer pairs with tf.keras.optimizers.schedules.LearningRateSchedule instead of a static learning
+        rate.
+
+        This code should function on CPU, GPU, and TPU.
+
+        Example:
+
+        ```python
+
+        model = get_model()
+
+        opt1 = tf.keras.optimizers.Adam(learning_rate=1e-4)
+        opt2 = tf.keras.optimizers.Adam(learning_rate=1e-2)
+
+        opt_layer_pairs = [(opt1, model.layers[0]),
+                            (opt2, model.layers[1:])]
+
+        loss = tf.keras.losses.MSE
+        optimizer = MultiOpt(opt_layer_pairs)
+
+        model.compile(optimizer=optimizer, loss = loss)
+
+        model.fit(x,y)
+
+        ```
+
+
+        """
+
+        super(MultiOpt, self).__init__(name, **kwargs)
+
+        if optimizer_specs is None and optimizer_layer_pairs is not None:
+            self.optimizer_specs = [
+                self.create_optimizer_spec(opt, layer)
+                for opt, layer in optimizer_layer_pairs
+            ]
+
+        elif optimizer_specs is not None and optimizer_layer_pairs is None:
+            self.optimizer_specs = optimizer_specs
+
+        else:
+            raise RuntimeError(
+                "You must specify either an list of optimizer_layer_pairs or a list of optimizer_specs"
+            )
+
+        self.initialized_optimizer_specs = [
+            self.initialize_from_optimizer_spec(spec) for spec in self.optimizer_specs
+        ]
+
+        self.lr = self.initialized_optimizer_specs[0]["optimizer"].lr
+
+    def apply_gradients(self, grads_and_vars, name=None):
+        """
+        Wrapped Gradient Apply method. Returns a list of tf ops to be executed.
+        """
+
+        for spec in self.optimizer_specs:
+            spec["gv"] = []
+
+        for grad, var in tuple(grads_and_vars):
+            for spec in self.optimizer_specs:
+                for weight in spec["weights"]:
+                    if var.name == weight.name:
+                        spec["gv"].append((grad, var))
+
+        return [
+            spec["optimizer"].apply_gradients(spec["gv"])
+            for spec in self.optimizer_specs
+        ]
+
+    def get_config(self):
+        config = super(MultiOpt, self).get_config()
+        config.update({"optimizer_specs": self.optimizer_specs})
+        return config
+
+    @classmethod
+    def initialize_from_optimizer_spec(cls, optimizer_spec):
+        optimizer_spec["optimizer"] = optimizer_spec["optimizer_class"].from_config(
+            optimizer_spec["optimizer_config"]
+        )
+        return optimizer_spec
+
+    @classmethod
+    def create_optimizer_spec(cls, optimizer_instance, layer):
+        if type(layer) == list:
+            weights = [
+                FakeVar(var.name) for sublayer in layer for var in sublayer.weights
+            ]
+        else:
+            weights = [FakeVar(var.name) for var in layer.weights]
```

Comment:
I'm not sure if that would make a difference because the wrapper calls the apply function of the wrapped optimizers. I could try it, but I believe that the wrapped optimizers have their own way of handling trainable vs non trainable weights in their apply step. I'm guessing that a normal opt receives [weights] when running model.fit so either the model.fit only passes trainable weights or the opt.apply filters them out somehow. If that's the case, then it would be unnecessary to filter the weights twice.

Reply:
The optimizer will not distinguish if the variable is trainable or not. It will create slots and compute gradient w.r.t. all variables passing to `apply_gradients`.

https://colab.research.google.com/drive/1NOigO7WuGdFKTKD4XovqD52SMrSYzI99?usp=sharing

Reply:
I think maybe that's intended behavior when using the optimizer outside of the model.fit loop? When used in the model.fit loop, the multi opt correctly ignores layers without the trainable attribute. 

I would prefer if the multi optimizer acts solely as a wrapper and also passes forward all bugs/features of the wrapped optimizer.

Reply:
link demonstrating layer trainable false works for multi optimizer [here](https://colab.research.google.com/drive/1NKAsrjj1qcls6p8S5dvJT1PkdC2Ieazc?usp=sharing)

Reply:
Got it. It looks good to me if that won't cause any errors.